### ISSUE TITLE
Error on IE 11: Object doesn't support property or method 'slice'

### ISSUE THREAD

**Issue by Sampadaak2613** [Hopding/pdf-lib#277]
Hi there,

I am using the pdf-lib library in SPFX list view command set to edit the footer of a PDF using drawtext. It works fine in google chrome browser. It is not working in IE 11. 
Below is the details of the same.

Reference Code : 

```js
import "@pnp/polyfill-ie11";
import 'es6-promise/auto';
import 'whatwg-fetch';
//import 'node_modules\core-js\es6\array.js';
//import 'node_modules\es6-map\implement.js'
/*import 'es6-promise';
import 'whatwg-fetch';
import 'polyfill-io-feature-detection';
import "core-js/stable";
require("regenerator-runtime/runtime");
require("@babel/polyfill");
require('es5-shim');
require('es6-shim');
require('array.from');*/

import {
    override
} from '@microsoft/decorators';
import {
    Log
} from '@microsoft/sp-core-library';
import {
    BaseListViewCommandSet,
    Command,
    IListViewCommandSetListViewUpdatedParameters,
    IListViewCommandSetExecuteEventParameters
} from '@microsoft/sp-listview-extensibility';
import {
    Dialog
} from '@microsoft/sp-dialog';

import * as strings from 'EditPdfFooterCommandSetStrings';

/* Custom Installed Modules */
//import * as jQuery from 'jquery';
//require('polyfillmin');
//require('jspdfmin');
//require('jspdfdebug');
import {
    degrees,
    PDFDocument,
    rgb,
    StandardFonts
} from 'pdf-lib';

export interface IEditPdfFooterCommandSetProperties {
    // This is an example; replace with your own properties
    sampleTextOne: string;
    sampleTextTwo: string;
}

const LOG_SOURCE: string = 'EditPdfFooterCommandSet';

export default class EditPdfFooterCommandSet extends BaseListViewCommandSet < IEditPdfFooterCommandSetProperties > {

    @override
    public onInit(): Promise < void > {
        Log.info(LOG_SOURCE, 'Initialized EditPdfFooterCommandSet');
        return Promise.resolve();

    }

    @override
    public onListViewUpdated(event: IListViewCommandSetListViewUpdatedParameters): void {
        const compareOneCommand: Command = this.tryGetCommand('COMMAND_1');
        if (compareOneCommand) {
            compareOneCommand.visible = event.selectedRows.length === 1;
        }
    }

    @override
    public async onExecute(event: IListViewCommandSetExecuteEventParameters): Promise < void > {
        switch (event.itemId) {
            case 'COMMAND_1':

                //Start - Using PDF-LIB
                var pdfUrl = "URL Of Sharepoint Document Library/Demo_PDF.pdf";
                const existingPdfBytes = await fetch(pdfUrl).then(res => res.arrayBuffer())
                const pdfDoc = await PDFDocument.load(existingPdfBytes) //Getting issue on this line of code
                const helveticaFont = await pdfDoc.embedFont(StandardFonts.Helvetica)
                const pages = pdfDoc.getPages()
                const firstPage = pages[0]
                const {
                    width, height
                } = firstPage.getSize()
                firstPage.drawText('This text was added with JavaScript!', {
                    x: 5,
                    y: height / 2 + 300,
                    size: 50,
                    font: helveticaFont,
                    color: rgb(0.95, 0.1, 0.1),
                    rotate: degrees(-45),
                })
                //Get pdf in Base64 format
                const pdfBytes = await pdfDoc.saveAsBase64({
                    dataUri: true
                });
                let pdfWindow = window.open("");
                pdfWindow.document.write("<iframe  width='100%' id='printf' name='printf' height='100%' src='" + encodeURI(pdfBytes) + "'></iframe>")
                //End - Using PDF-LIB
                Dialog.alert(`${this.properties.sampleTextOne}`);
                break;
            case 'COMMAND_2':
                Dialog.alert(`${this.properties.sampleTextTwo}`);
                break;
            default:
                throw new Error('Unknown command');
        }
    }
}
```

----------------------------------------------------------

To fix this issue on IE 11, I have installed below modules but no luck.

```
npm i polyfill-io-feature-detection
npm install whatwg-fetch
npm install es6-promise
npm install es6-shim
npm install es5-shim
npm install array.from
npm install --save @babel/polyfill
npm i promise-polyfill
npm install --save @pnp/polyfill-ie11 
```

------------------------------------------------------------------

Getting below error on IE 11

![image](https://user-images.githubusercontent.com/58074040/71340092-ace33d00-257b-11ea-8e12-daef6b201011.png)

![image](https://user-images.githubusercontent.com/58074040/71340076-a228a800-257b-11ea-9540-ee181227b71e.png)

![image](https://user-images.githubusercontent.com/58074040/71340084-a81e8900-257b-11ea-830d-545060675b51.png)

./node_modules/pdf-lib/es/core/parser/ByteStream.js
./node_modules/pdf-lib/es/core/parser/PDFParser.js

![image](https://user-images.githubusercontent.com/58074040/71341060-c639b880-257e-11ea-933e-067381f6bd96.png)

------------------------------------------------------------------

Please find attached document for your reference.

[PDF-LIB IE 11 Issue.docx](https://github.com/Hopding/pdf-lib/files/3993876/PDF-LIB.IE.11.Issue.docx)

**Comment by Hopding** [comment]
Hello @Sampadaak2613. 

Many JavaScript libraries are built for modern browsers and will not work out of the box on outdated browsers like IE11. This is the case with pdf-lib. If you want to get pdf-lib working on IE11, you'll probably have to polyfill the missing features.

I do not have a way to test on IE11. So all I can do to help you with this is look at the error you've shared and suggest that you add a polyfill for the missing method. Specifically, you are encountering the following error:
```
TypeError: Object doesn't support property or method 'slice'
```
This error is being thrown from the following method of [`ByteStream`](https://github.com/Hopding/pdf-lib/blob/645a5307a35ea36994bbf9925197132f9c3aaccd/src/core/parser/ByteStream.ts):

https://github.com/Hopding/pdf-lib/blob/645a5307a35ea36994bbf9925197132f9c3aaccd/src/core/parser/ByteStream.ts#L67-L69

The `this.bytes` being sliced here is a `Uint8Array`. So from all this, we can infer that IE11 does not provide an implementation for the [`Uint8Array.slice`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/slice) method. The only viable solution to move past this error is to polyfill the method.

It's not clear to me whether you had any success using the polyfill I suggested in [#274](https://github.com/Hopding/pdf-lib/issues/274#issuecomment-568269650). So I'll suggest that you try it again. I would advise placing it just above the class declaration in the code snippet you shared. For example:
```diff
const LOG_SOURCE: string = 'EditPdfFooterCommandSet';

+if (!Uint8Array.prototype.slice) {
+  Object.defineProperty(Uint8Array.prototype, 'slice', {
+    value: function (begin, end)
+     {
+        return new Uint8Array(Array.prototype.slice.call(this, begin, end));
+     }
+  });
+}

export default class EditPdfFooterCommandSet extends BaseListViewCommandSet < IEditPdfFooterCommandSetProperties > {

    @override
    public onInit(): Promise < void > {
        Log.info(LOG_SOURCE, 'Initialized EditPdfFooterCommandSet');
        return Promise.resolve();
    }
```

**Comment by jotae1e** [comment]
Thanks a lot.  

I tried a lot of polyfills, but yours did the job.

Thanks a lot.

**Comment by Sampadaak2613** [comment]
Hi Andrew,

Thanks for your reply.
As per my understanding, we need to use polyfills of those actions which
are not supported by IE 11. Please confirm.

You have given polyfill of Slice method.
I will add it on the top of my code.
I was trying to edit Slice method from ByteStream.js but no luck.

I will let you know if any issues.
Once again, thanks a lot for your help.


Regards,
Sampada

**Comment by Hopding** [comment]
@Sampadaak2613 

> As per my understanding, we need to use polyfills of those actions which are not supported by IE 11. Please confirm.

Yes. This is correct.

**Comment by Sampadaak2613** [comment]
Hi,

Thanks a lot for your help.
As suggested by you, I have added below polyfills in my code after `const LOG_SOURCE: string = 'EditPdfFooterCommandSet';` and it worked.

`
//Polyfill for slice 
//https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/slice
if (!Uint8Array.prototype.slice) {
  Object.defineProperty(Uint8Array.prototype, 'slice', {
    value: function (begin, end)
     {
        return new Uint8Array(Array.prototype.slice.call(this, begin, end));
     }
  });
}


//Polyfill for codePointAt
//http://xahlee.info/js/js_String.prototype.codePointAt.html
if ( ! String.prototype.codePointAt ) {

  String.prototype.codePointAt = function fixedCharCodeAt (idx) {
      // 2017-02-24 from https://developer.mozilla.org/en-US/docs/JavaScript/Reference/Global_Objects/String/charCodeAt
      // 2017-02-24 xah lee, made to be compatible with es2015 codePointAt
      // ex. fixedCharCodeAt ('\uD800\uDC00', 0); // 65536
      // ex. fixedCharCodeAt ('\uD800\uDC00', 1); // 65536
      idx = idx || 0
      var code = this.charCodeAt(idx)
      var hi, low
      if (0xD800 <= code && code <= 0xDBFF) { // High surrogate (could change last hex to 0xDB7F to treat high private surrogates as single characters)
          hi = code
          low = this.charCodeAt(idx+1)
          if (isNaN(low)) {
              throw 'High surrogate not followed by low surrogate in fixedCharCodeAt()'
          }
          return ((hi - 0xD800) * 0x400) + (low - 0xDC00) + 0x10000
      }
      if (0xDC00 <= code && code <= 0xDFFF) { // Low surrogate
          // We return false to allow loops to skip this iteration since should have already handled high surrogate above in the previous iteration
          return false
          /*hi = this.charCodeAt(idx-1)
            low = code
            return ((hi - 0xD800) * 0x400) + (low - 0xDC00) + 0x10000;*/
      }
      return code
  };
}

//Polyfill for Includes
//https://stackoverflow.com/questions/31221341/ie-does-not-support-includes-method
if (!Array.prototype.includes) {
  Object.defineProperty(Array.prototype, "includes", {
    enumerable: false,
    value: function(obj) {
        var newArr = this.filter(function(el) {
          return el == obj;
        });
        return newArr.length > 0;
      }
  });
}

`

But now, after editing the pdf, I am rendering it to Iframe in IE 11, but nothing dispplayed on the browser. I have converted Pdf Base64 format to Blob format and tried to display it in IE but no luck.
I have also used atob, btoa, Object, Embed but nothing happens in IE.

`
//Get pdf in Base64 format
                        const pdfBytes = await pdfDoc.saveAsBase64({
                            dataUri: true
                        });
                        let pdfWindow =window.open("");
                        
                        pdfWindow.document.write("<iframe width='100%' id='printf' name='printf' height='100%' src='" + encodeURI(pdfBytes) + "'></iframe>")                       

`

**Comment by Hopding** [comment]
@Sampadaak2613 

This might just be an issue with IE11. I did a bit of googling and it looks like others have also has issues rendering PDFs in `iframes` on IE11. But I can't say for sure, because I cannot test it myself.

However, the code snippet you shared also doesn't look quite right. I'd suggest taking a look at some of the [`web` example tests](https://github.com/Hopding/pdf-lib/tree/master/apps/web) to see how to properly render a PDF in a `iframe` on the browser. In particular, [`test1.html`](https://github.com/Hopding/pdf-lib/blob/cb0bca1561b8ceb26e099c9a37bc34c2d946fbcf/apps/web/test1.html) shows how to render the raw bytes produced by `pdfDoc.save()`. And [`test3.html`](https://github.com/Hopding/pdf-lib/blob/cb0bca1561b8ceb26e099c9a37bc34c2d946fbcf/apps/web/test3.html) shows how to render the base64 string produced by `pdfDoc.saveAsBase64({ dataUri: true })`. 

The following example (available on the project website: https://pdf-lib.js.org) also demonstrates how to render a base64 string:
```html
<html>
  <head>
    <meta charset="utf-8" />
    <script src="https://unpkg.com/pdf-lib"></script>
  </head>

  <body>
    <iframe id="pdf" style="width: 100%; height: 100%;"></iframe>
  </body>

  <script>
    createPdf();
    async function createPdf() {
      const pdfDoc = await PDFLib.PDFDocument.create();
      const page = pdfDoc.addPage([350, 400]);
      page.moveTo(110, 200);
      page.drawText('Hello World!');
      const pdfDataUri = await pdfDoc.saveAsBase64({ dataUri: true });
      document.getElementById('pdf').src = pdfDataUri;
    }
  </script>
</html>
```

Finally, if you are still unable to get things working after trying what I mentioned above, it's probably just an IE11 issue. In which case your best bet is probably to avoid rendering the PDF in an `iframe` entirely. The most flexible alternative that I'm aware of is to render your PDFs with Mozilla's [`pdf.js` viewer](https://github.com/mozilla/pdf.js). Here are a few tutorials that I pulled from a google search that should help you get started:
* https://www.sitepoint.com/custom-pdf-rendering/
* https://mozilla.github.io/pdf.js/examples/
* https://pspdfkit.com/blog/2018/render-pdfs-in-the-browser-with-pdf-js/
* https://gist.github.com/fcingolani/3300351

If you don't want to (or for some reason can't) use `pdf.js`, the only other option I know of is to render your PDFs with Google or Microsoft's server-side rendering viewers: https://gist.github.com/tzmartin/1cf85dc3d975f94cfddc04bc0dd399be. Of course, this is a pretty inflexible approach and also requires you to send Google/Microsoft your documents, which isn't ideal (and makes this option a nonstarter for many people).

I hope this helps. Good luck with your project!

**Comment by Sampadaak2613** [comment]
Hi there,

Thanks for your reply but I am still not able to find the solution. 
I converted PDF Base 64 format to Blob and tried it to render to IE but it doesn't worked but it worked in Chrome.
I tried with iFrame, Object, Embed but still no luck
I also tried 1 more approach but it is downloding the pdf instead of opening it in browser new tab.

`
            //Get pdf in Base64 format
            const pdfBytes1 = await pdfDoc.saveAsBase64({
                dataUri: true
            });

            //Remove data:application/pdf;base64, from pdfbytes string
            var res = pdfBytes1.split("data:application/pdf;base64,")[1];
            var bytes = atob(res);
            //Below code downloads the pdf file instead of opening the browser            
            if (window.navigator && window.navigator.msSaveOrOpenBlob) {
                var byteCharacters = atob(res);
                var byteNumbers = new Array(byteCharacters.length);
                for (var i = 0; i < byteCharacters.length; i++) {
                    byteNumbers[i] = byteCharacters.charCodeAt(i);
                }
                var byteArray = new Uint8Array(byteNumbers);
                var blob = new Blob([byteArray], {
                    type: 'application/pdf'
                });
                window.navigator.msSaveOrOpenBlob(blob, "Test.pdf");
            } else {
                // Directly use base 64 encoded data for rest browsers (not IE)
                var base64EncodedPDF = res;
                var dataURI = pdfBytes1;
                window.open(dataURI, '_blank');
            }

            //Converting base64 format to blob
            const pdfBytes1 = await pdfDoc.save()
            var byteArray = new Uint8Array(pdfBytes);
            var blob = new Blob([byteArray], {
                type: 'application/pdf'
            });
            var tab = window.open();
            var objectUrl = URL.createObjectURL(blob);

            if (window.navigator && window.navigator.msSaveOrOpenBlob) {
                //Trying to load it in iFrame
                var iframe = document.createElement("iframe");
                iframe.src = objectUrl;
                iframe.setAttribute("style", "width:100%;height:100%");
                tab.document.appendChild(iframe);

                // Second approach. 
                var mywindow = window.open('about:blank', 'Print', 'height=800,width=900');
                mywindow.document.write('<meta http-equiv="X-UA-Compatible" content="IE=9; IE=8; IE=7" /> ')
                mywindow.document.write('<!DOCTYPE html><head>');
                mywindow.document.write('<link rel="stylesheet" href="./css/scr.css" type="text/css" />');
                mywindow.document.write('</head><body><iFrame  src=' + objectUrl + '></iFrame>');
                //add logos and legend here.
                mywindow.document.write('<meta http-equiv="X-UA-Compatible" content="IE=9; IE=8; IE=7" /> ')
                //mywindow.document.body.appendChild(iframe);        
                mywindow.document.write('</div></body></html>');
                mywindow.document.write('<meta http-equiv="X-UA-Compatible" content="IE=9; IE=8; IE=7" /> ')

                //window.navigator.msSaveOrOpenBlob(blob);

            } else {
                var objectUrl = URL.createObjectURL(blob);
                window.open(objectUrl);
            }
            //Below code opens blank iFrame page in IE 
            //let pdfWindow = window.open("");
            // pdfWindow.document.write("<iframe  width='100%' id='printf' name='printf' height='100%' src='"+encodeURI(pdfBytes1)+"'></iframe>");
`

Let me know your inputs on it.

**Comment by Sampadaak2613** [comment]
Hi there,

I have tried with pdf.js. Its working in Chrome but not in IE.
Below is the code snippet.


`
import {getDocument} from 'pdfjs-dist' ;
var loadingTask = getDocument({data: bytes});
                    loadingTask.promise.then(function(pdf) {
                      console.log('PDF loaded');
                      
                      // Fetch the first page
                      var pageNumber = 1;
                      pdf.getPage(pageNumber).then(function(page) {
                        console.log('Page loaded');
                        
                        var scale = 1.5;
                        var viewport = page.getViewport({scale: scale});
                    
                        // Prepare canvas using PDF page dimensions
                        document.write('<meta http-equiv="X-UA-Compatible" content="IE=9; IE=8; IE=7" /> ')
                        document.write('<!DOCTYPE html><head>');
                        document.write('</head><body><canvas id="the-canvas" ></canvas>'); 
                        document.write('</body></html>');                     
                        var canvas : any = document.getElementById('the-canvas');
                        var context = canvas.getContext('2d');
                        canvas.height = viewport.height;
                        canvas.width = viewport.width;
                    
                        // Render PDF page into canvas context
                        var renderContext = {
                          canvasContext: context,
                          viewport: viewport
                        };
                        var renderTask = page.render(renderContext);
                        renderTask.promise.then(function () {
                          console.log('Page rendered');
                        });
                      });
                    }, function (reason) {
                      // PDF loading error
                      console.error(reason);
                    });
`

Now I am trying easyPDF.js https://www.jqueryscript.net/other/Base64-PDF-Viewer-Easy-PDF.html

**Comment by Hopding** [comment]
@Sampadaak2613 I'm afraid there's not much I can do to help you other than what I recommended in https://github.com/Hopding/pdf-lib/issues/277#issuecomment-569075645. All the approaches I suggest there work on all the modern browsers (Chrome, Firefox, and Safari) I have access to. 

If you're unable to get `pdf.js` working, I'm afraid I can't help you with that. If you're unable to find a solution on the internet, I'd suggest you open an issue on the GitHub repo for `pdf.js` (https://github.com/mozilla/pdf.js/). Or you could ask a question on StackOverflow about it.

One last thing: If you copy the HTML example I provided and paste it into an HTML file (e.g. `test.html`), and open that HTML file in IE, exactly what happens?

**Comment by mcshaz** [comment]
I am  just posting here to confirm that this library can definately work with webpack/babel/core-js@3. I had to direct babel to transpile dependencies for `pdf-lib` (by default, node_modules are not transpiled in most build processes). I was using Vue-cli, and so it was as simple as adding:
```javascript
module.exports = {
  transpileDependencies: ['pdf-lib'],
````

I also ran into trouble with IE11 intl.dateTimeFormat implementation, where for some reason when you split the output of any of the related formatting functions (e.g. `new Date().toLocaleDateString().split('')`) (as is done under the hood when pdf-lib adds text to the content stream), multiple charCode 8206 (left to right mark (LRM)) are in the resulting string. These must be stripped out or the pdf-lib function `encodeUnicodeCodePoint` will throw;
